### PR TITLE
Make the release tagging process more rigorous

### PR DESCRIPTION
**Pull request by jskeet**
- Fetch the github master branch head commit, and use this consistently
- Local repo must be at the same commit
- Local repo must not have changes apis/apis.json
- Project references for packages to release must not "escape" that
  set of packages. (If package X has a project reference to package Y,
  both X and Y must be released together.)

### CODE REVIEW

File: tools/Google.Cloud.Tools.TagReleases/Program.cs [hunk 1]
```diff
@@ -45,70 +52,133 @@ private static int Main(string[] args)
...
+            return allApis.Except(noChange).ToList();
+        }
+
+        /// <summary>
+        /// Validates the set of changes.
+        /// </summary>
+        /// <remarks>
+        /// <para>Current checks:</para>
+        /// <para>
+        /// Project references (in produuction code) are okay so long as all the targets of the references
```
Comment (chrisdunelm) [inline tools/Google.Cloud.Tools.TagReleases/Program.cs]: nit: prod*uu*ction

File: tools/Google.Cloud.Tools.TagReleases/Program.cs [hunk 2]
```diff
@@ -45,70 +52,133 @@ private static int Main(string[] args)
...
+        /// are also going to be released. If this is not the case, the dependencies within the target could be different
+        /// to the ones in the public package version, causing a dependency issue in the package we're about to publish.
+        /// (This caused issue 1280 for example.)
```
Comment (chrisdunelm) [inline tools/Google.Cloud.Tools.TagReleases/Program.cs]: nit: issue *#*1280

Comment (jskeet) [inline tools/Google.Cloud.Tools.TagReleases/Program.cs]: I don't think that'll do anything within the source code, to be honest... it's not like it's a github commit. Happy to do it anyway though.

**Comment by jskeet** [comment]
Thanks - will amend and merge in a bit.

**Pull request merged**